lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp

```diff
@@ -4634,6 +4634,211 @@ class DecomposeAtenTypeAsOp : public OpRewritePattern<AtenTypeAsOp> {
 };
 } // namespace
 
+// AtenIndexTensorOp
+namespace {
+// Strictly speaking, this pattern is not a decomposition. The goal of this
+// pattern is to eliminate none index in aten.Index.Tensor's `indices` param for
+// the ease of various backend. The detailed steps are:
+//    1. reorder input tensor so that the non-none index appears at adjacent
+//    positions.
+//    2. manually generate index tensor with some ops like iota, to replace the
+//    none index in `indices`
+//    3. replace the old aten.Index.Tensor with a new
+//    aten.Index.Tensor_hacked_twin.
+class DecomposeAtenIndexTensorOp : public OpRewritePattern<AtenIndexTensorOp> {
```

Comment:
We should remove the decomposition for `HackedTwin` -> `aten.index.Tensor`, to avoid a cycle of decompositions. Also, `AtenIndexTensorOp` should be added to

https://github.com/llvm/torch-mlir/blob/574450687d8361bfa055b9cc09e0bef91541d4c3/lib/Dialect/Torch/Transforms/LowerToBackendContract.cpp#L364

to make it clear to backends that they only need to worry about the hacked twin variant

Reply:
Removing decomposition for `aten.Index.Tensor_hacked_twin` op is reasonable. Since the pass only works when the indices of `aten.Index.Tensor` contains none value. If we mark `aten.Index.Tensor` illegal explicitly, we will fail to lower some `aten.Index.Tensor` ops.

Reply:
A possible solution may be that we fill the `indices` argument using the methods in this PR until it has the same length as the input_tensor's rank. But it will cause many e2e tests fail, because all backends only handle aten.Index.Tensor for now.

Reply:
> But if we mark aten.Index.Tensor illegal explicitly, we will fail to lower some aten.Index.Tensor ops, since the pass only works when the indices argument contains none value.

We should have the pattern also decompose into the hacked twin when there are no `None`s in the original `aten.Index.Tensor`. It would just be a renaming. We want to have at most one variant around at the end to make it easy for backends to know what needs handling.

> But it will cause many e2e tests fail, because all backends only handle aten.Index.Tensor for now.

Yeah, we need to update the backends to lower the hacked variant instead of `aten.Index.Tensor`. I think this would be straight forward. We just need to remove the logic that handles the `None` cases and replace every `AtenIndexTensorOp` with the hacked twin op